### PR TITLE
Uninstall support  for ovn

### LINKED ISSUE #1868: Cleanup OVN resources during uninstall
**What would you like to be added**:
Cleanup all OVN resources during uninstall if OVNKubernetes is the CNI

**Why is this needed**:
During install, networkplugin-syncer and route-agent add some resources to OVN. Need to make sure these are all cleaned up properly during an uninstall.

### PR DESCRIPTION
**Pull request by aswinsuryan**
Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

Fixes #1868

### CODE REVIEW

File: pkg/routeagent_driver/handlers/ovn/vsctl/vsctl.go [hunk 1]
```diff
@@ -70,6 +76,12 @@ func AddInternalPort(bridgeName, portName, macAddress string, mtu int) error {
 	return err
 }
 
+func DelInternalPort(bridgeName, portName string) error {
+	_, err := vsctlCmd("--may-exist", "del-port", bridgeName, portName)
```
Comment (sridhargaddam) [inline pkg/routeagent_driver/handlers/ovn/vsctl/vsctl.go]:
```suggestion
	_, err := vsctlCmd("--if-exists", "del-port", bridgeName, portName)
```

We should use "if-exists" during delete port.

```
Without  --if-exists,  attempting to delete a port that does not exist is an error.  With --if-exists,  attempting  to  delete  a port that does not exist has no effect.
```

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	"k8s.io/klog/v2"
+)
+
+func (ovn *SyncHandler) Stop(uninstall bool) error {
+	if !uninstall {
+		return nil
+	}
+
+	klog.Infof("Uninstalling Submariner changes from the node")
```
Comment (sridhargaddam) [inline pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: Network-plugin syncer changes are cluster wide and not specific to the node. Please amend the sentence accordingly.

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+	// Delete the submariner logical router, ports and flows
+	staleLRSRPred := func(item *nbdb.LogicalRouterStaticRoute) bool {
+		return item.OutputPort != nil && *item.OutputPort == submarinerUpstreamRPort
+	}
+
+	err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(ovn.nbdb, submarinerLogicalRouter, staleLRSRPred)
+	if err != nil {
+		klog.Errorf("failed to delete ovn static routes for port: %s due to : %v", submarinerUpstreamRPort, err)
```
Comment (sridhargaddam) [inline pkg/networkplugin-syncer/handlers/ovn/uninstall.go]:
```suggestion
		klog.Errorf("Failed to delete ovn static routes for port: %s due to : %v", submarinerUpstreamRPort, err)
```

Although we do not enforce this. AFAICT, the common pattern we follow is to use capital letter when printing the error message and use small letters when returning the error.

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		klog.Errorf("failed to delete router ports from submariner logical router due to %v", err)
+	}
+
+	err = libovsdbops.DeleteLogicalRouter(ovn.nbdb, &subLogicalRouter)
+	if err != nil {
+		klog.Errorf("failed to delete submariner logical router %v", err)
+	}
+
+	// Delete the ports and flows
```
Comment (sridhargaddam) [inline pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: Is it port or policies?

File: pkg/routeagent_driver/handlers/ovn/uninstall.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/submariner-io/submariner/pkg/routeagent_driver/handlers/ovn/vsctl"
+	"k8s.io/klog"
+)
+
+func (ovn *Handler) Stop(uninstall bool) error {
+	if !uninstall {
+		return nil
+	}
+
+	klog.Infof("Uninstalling Submariner changes from the node")
```
Comment (sridhargaddam) [inline pkg/routeagent_driver/handlers/ovn/uninstall.go]: We also program some iptable rules in the filter table to clamp tcpMss. Can you please cleanup those rules as well.

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Thanks for the PR @aswinsuryan

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go (hunk 2 above)
Comment (tpantelis) [inline pkg/networkplugin-syncer/handlers/ovn/uninstall.go]:
```suggestion
	klog.Infof("Uninstalling OVN components")
```

File: pkg/routeagent_driver/handlers/ovn/uninstall.go (hunk 5 above)
Comment (tpantelis) [inline pkg/routeagent_driver/handlers/ovn/uninstall.go]:
```suggestion
	klog.Infof("Uninstalling OVN components from the node")
```

File: pkg/routeagent_driver/handlers/ovn/uninstall.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+	err := vsctl.DelInternalPort(ovnK8sSubmarinerBridge, ovnK8sSubmarinerInterface)
+	if err != nil {
+		klog.Errorf("error deleting Submariner port %q due to %v", ovnK8sSubmarinerInterface, err)
```
Comment (tpantelis) [inline pkg/routeagent_driver/handlers/ovn/uninstall.go]: As mentioned above, all log messages should be capitalized. 

```suggestion
		klog.Errorf("Error deleting Submariner port %q due to %v", ovnK8sSubmarinerInterface, err)
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go (hunk 4 above)
Comment (aswinsuryan) [reply pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: we delete ports and policies. Now changed  flows to policies

File: pkg/routeagent_driver/handlers/ovn/uninstall.go (hunk 5 above)
Comment (aswinsuryan) [reply pkg/routeagent_driver/handlers/ovn/uninstall.go]: it is cleaned up when we call cleanupGatewayDataplane which calls cleanupForwardingIptables

[1]https://github.com/submariner-io/submariner/blob/938a049794cfb02b6ce1e5dde7932d5e93382e51/pkg/routeagent_driver/handlers/ovn/gateway_dataplane.go#L176

File: pkg/routeagent_driver/handlers/ovn/uninstall.go (hunk 6 above)
Comment (aswinsuryan) [reply pkg/routeagent_driver/handlers/ovn/uninstall.go]: Done.

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: Done.

File: pkg/routeagent_driver/handlers/ovn/uninstall.go (hunk 5 above)
Comment (aswinsuryan) [reply pkg/routeagent_driver/handlers/ovn/uninstall.go]: Done.

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go (hunk 3 above)
Comment (aswinsuryan) [reply pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: Done.

File: pkg/networkplugin-syncer/handlers/ovn/uninstall.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/networkplugin-syncer/handlers/ovn/uninstall.go]: Done.

File: pkg/routeagent_driver/handlers/ovn/vsctl/vsctl.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/routeagent_driver/handlers/ovn/vsctl/vsctl.go]: Done.

**Review by tpantelis — APPROVED** [review]

**Review by sridhargaddam — APPROVED** [review]

**Comment by nyechiel** [comment]
@sridhargaddam @aswinsuryan should we merge this over the broken kind setup, or do we want to wait?

**Comment by aswinsuryan** [comment]
> @sridhargaddam @aswinsuryan should we merge this over the broken kind setup, or do we want to wait?

@nyechiel the changes are only in uninstall path, and the broken jobs do not verify that, so I think it should be safe to merge

**Comment by nyechiel** [comment]
@aswinsuryan please backport to `release-0.13` and link the PR to https://github.com/submariner-io/releases/issues/418

**Pull request merged**